src/rust/engine/process_execution/src/speculate.rs

```diff
@@ -33,19 +34,42 @@ impl SpeculatingCommandRunner {
     req: MultiPlatformExecuteProcessRequest,
     context: Context,
   ) -> BoxFuture<FallibleExecuteProcessResult, String> {
-    let command_runner = self.clone();
-    let req_2 = req.clone();
     let delay = Delay::new(Instant::now() + self.speculation_timeout);
+    let req2 = req.clone();
     self
       .primary
       .run(req, context.clone())
-      .select(delay.then(move |_| command_runner.secondary.run(req_2, context)))
+      .select2({
+        let command_runner = self.clone();
+        delay.then(move |_| command_runner.secondary.run(req2, context))
+      })
       .then(|raced_result| match raced_result {
-        Ok((successful_res, _outstanding_req)) => {
-          ok::<FallibleExecuteProcessResult, String>(successful_res).to_boxed()
+        Ok(either_success) => {
+          // split take out the homogeneous success type for either primary or
+          // secondary successes.
+          ok::<FallibleExecuteProcessResult, String>(either_success.split().0).to_boxed()
+        }
+        Err(Either::A((failed_primary_res, _))) => {
+          debug!("primary request FAILED, aborting");
+          err::<FallibleExecuteProcessResult, String>(failed_primary_res).to_boxed()
         }
-        Err((failed_res, _outstanding_req)) => {
-          err::<FallibleExecuteProcessResult, String>(failed_res).to_boxed()
+        // We handle the case of the secondary failing specially. We only want to show
+        // a failure to the user if the primary execution source fails. This maintains
+        // feel between speculation on and off states.
+        Err(Either::B((_failed_secondary_res, outstanding_primary_request))) => {
+          debug!("secondary request FAILED, waiting for primary!");
+          outstanding_primary_request
+            .then(|primary_result| match primary_result {
```

Comment:
The return type of this closure is consistent across the branches, so we don't need the boxing or the future-result type ascriptions and stuff. This could actually just be:
```rust
            .then(|primary_result| match primary_result {
              Ok(primary_success) => {
                debug!("primary request eventually SUCCEEDED after secondary failed");
                Ok(primary_success)
              },
                Err(primary_failure) => {
                debug!("primary request eventually FAILED after secondary failed");
                Err(primary_failure)
              })
```

or even simpler:
```rust
.then(|primary_result| {
  if primary_result.is_ok() {
    debug!("primary request eventually SUCCEEDED after secondary failed");
  } else {
    debug!("primary request eventually FAILED after secondary failed");
  }
  primary_result
})

Reply:
This is helpful feedback thanks.